### ISSUE TITLE
Items from MMOItems in recipes stopped working

### ISSUE THREAD

**Issue by ShadySt** [WolfyScript/CustomCrafting#367]
### What happened?

I was using CC 4.16.6.4 on 1.19.3 server with MMOItems for a long time. 
Now I decided to update to 1.20.4 but discovered, that relevant versions of CC and WU having big troubles with MMOItems. 
If I'm trying to create new recipe with MMOItems in it, using ingame plugin GUI, I'm getting warns in console and chat message "error saving recipe! Cannot save an empty recipe" on recipe save.

If I'm trying to load my old recipes created on previous version, I get a lot of console messages, most useful of which I think are:
```
[13:17:34] [ForkJoinPool-5-worker-2/WARN]: [CustomCrafting] [LOCAL] [Error] Invalid Recipe: "customcrafting:foodsimple/shashlik_pork": GenerateLoreEvent may only be triggered synchronously. (through reference chain: me.wolfyscript.customcrafting.recipes.CustomRecipeCampfire["source"]->me.wolfyscript.customcrafting.recipes.items.Ingredient["items"]->java.util.ArrayList[0])
[13:17:34] [ForkJoinPool-5-worker-2/WARN]: [CustomCrafting] [LOCAL] [Error]   Caused by: GenerateLoreEvent may only be triggered synchronously.
[13:17:34] [ForkJoinPool-5-worker-2/WARN]: [CustomCrafting] ------------------[StackTrace]-------------------
[13:17:34] [ForkJoinPool-5-worker-2/WARN]: me.wolfyscript.lib.com.fasterxml.jackson.databind.JsonMappingException: GenerateLoreEvent may only be triggered synchronously. (through reference chain: me.wolfyscript.customcrafting.recipes.CustomRecipeCampfire["source"]->me.wolfyscript.customcrafting.recipes.items.Ingredient["items"]->java.util.ArrayList[0])
[13:17:34] [ForkJoinPool-5-worker-2/WARN]: 	at wolfyutils-spigot-4.16.15.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:392)
```

I searched Discord for similar problems and found this:
> WolfyScript — 10/03/2024 12:10
> ok looks like an incompatibility with the MMOItems GenerateLoreEvent. I need to figure out a solution for that, but that may take a while. For now you can't use MMOItems inside recipes.

Nothing related to this problem found in changelogs or anywhere else later.
 

### How can the bug be reproduced?

Try to create or load any recipe containing any item from MMOItems as ingredient or result. 

### Expected Behaviour?

Successful recipe creation, same as in CC 4.16.6.4 

### Plugin version

CustomCrafting 4.16.9.1, WolfyUtils 4.17-beta1

### Server software

Paper 1.20.4

### Relevant log output

```shell
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: [CustomCrafting] [LOCAL] [Error] Invalid Recipe: "customcrafting:test/test1": GenerateLoreEvent may only be triggered synchronously. (through reference chain: me.wolfyscript.customcrafting.recipes.CraftingRecipeShaped["ingredients"]->java.util.LinkedHashMap["B"]->me.wolfyscript.customcrafting.recipes.items.Ingredient["items"]->java.util.ArrayList[0])
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: [CustomCrafting] [LOCAL] [Error]   Caused by: GenerateLoreEvent may only be triggered synchronously.
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: [CustomCrafting] ------------------[StackTrace]-------------------
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: me.wolfyscript.lib.com.fasterxml.jackson.databind.JsonMappingException: GenerateLoreEvent may only be triggered synchronously. (through reference chain: me.wolfyscript.customcrafting.recipes.CraftingRecipeShaped["ingredients"]->java.util.LinkedHashMap["B"]->me.wolfyscript.customcrafting.recipes.items.Ingredient["items"]->java.util.ArrayList[0])
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:392)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:363)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.deser.std.CollectionDeserializer._deserializeFromArray(CollectionDeserializer.java:371)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:244)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//com.wolfyscript.jackson.dataformat.hocon.deserialization.ModifiedCollectionDeserializer.deserialize(ModifiedCollectionDeserializer.java:31)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//com.wolfyscript.jackson.dataformat.hocon.deserialization.ModifiedCollectionDeserializer.deserialize(ModifiedCollectionDeserializer.java:15)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:542)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeWithErrorWrapping(BeanDeserializer.java:563)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:438)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1405)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:351)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:184)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.deser.std.MapDeserializer._readAndBind(MapDeserializer.java:547)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:440)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:32)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:542)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeWithErrorWrapping(BeanDeserializer.java:563)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:470)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1405)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:351)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeOther(BeanDeserializer.java:219)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:186)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer._deserializeTypedForId(AsPropertyTypeDeserializer.java:144)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer.deserializeTypedFromObject(AsPropertyTypeDeserializer.java:110)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.deser.AbstractDeserializer.deserializeWithType(AbstractDeserializer.java:263)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.deser.impl.TypeWrappedDeserializer.deserialize(TypeWrappedDeserializer.java:74)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:323)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ObjectReader._bindAndClose(ObjectReader.java:2051)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ObjectReader.readValue(ObjectReader.java:1599)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at customcrafting-spigot-4.16.9.1.jar//me.wolfyscript.customcrafting.handlers.LocalStorageLoader$NewDataLoader.lambda$loadRecipesInNamespace$2(LocalStorageLoader.java:483)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1395)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: Caused by: java.lang.IllegalStateException: GenerateLoreEvent may only be triggered synchronously.
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at io.papermc.paper.plugin.manager.PaperEventManager.callEvent(PaperEventManager.java:42)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at io.papermc.paper.plugin.manager.PaperPluginManagerImpl.callEvent(PaperPluginManagerImpl.java:126)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:615)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at MMOItems-6.10-20240415.020818-7.jar//net.Indyuce.mmoitems.api.item.build.ItemStackBuilder.buildNBT(ItemStackBuilder.java:218)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at MMOItems-6.10-20240415.020818-7.jar//net.Indyuce.mmoitems.api.item.build.ItemStackBuilder.buildNBT(ItemStackBuilder.java:114)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at MMOItems-6.10-20240415.020818-7.jar//net.Indyuce.mmoitems.api.item.build.ItemStackBuilder.buildSilently(ItemStackBuilder.java:266)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.utilities.compatibility.plugins.mmoitems.MMOItemsStackIdentifier.stack(MMOItemsStackIdentifier.java:61)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.utilities.api.inventory.custom_items.references.APIReference.convertToStackReference(APIReference.java:164)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//com.wolfyscript.utilities.bukkit.world.items.reference.StackReference$Deserializer.lambda$convert$1(StackReference.java:386)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at java.base/java.util.Optional.map(Optional.java:260)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//com.wolfyscript.utilities.bukkit.world.items.reference.StackReference$Deserializer.convert(StackReference.java:379)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//com.wolfyscript.utilities.bukkit.world.items.reference.StackReference$Deserializer.convert(StackReference.java:336)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.deser.std.StdNodeBasedDeserializer.deserialize(StdNodeBasedDeserializer.java:73)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	at wolfyutils-spigot-4.17-beta1.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.deser.std.CollectionDeserializer._deserializeFromArray(CollectionDeserializer.java:355)
[16:53:59] [ForkJoinPool-5-worker-1/WARN]: 	... 33 more
```

### Submit issue

- [X] read the Wiki and especially the FaQ, and made sure there is no fix to your issue,
- [X] searched for and ensured there isn't already an open issue regarding this,
- [X] are running the latest version of your server software (Spigot, Paper, etc.), and plugins (WolfyUtilities, CustomCrafting).

**Comment by WolfyScript** [comment]
> Nothing related to this problem found in changelogs or anywhere else later.
Yeah... kinda forgot about the issue after that...

It seems to affect the upgrade feature to the new recipe format.
The only way I could fix it, would be to disable the asynchronous recipe loading.

Unfortunately, setting the cores of the loader to 1 in the config still loads them async, just on a single thread.
I think I'll just add a separate config option to toggle between sync/async loading.

**Comment by JohnKillswitch** [comment]
same in latest CC build (4.16.9.1)

**Comment by ShadySt** [comment]
Is there any update on this issue? I updated CC to 4.16.9.1 + WolfyUtils 4.17-beta.3.
MMOItems 6.10 
Still can't create any recipe but with slightly different error.

`[14:53:09 WARN]: me.wolfyscript.lib.com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Direct self-reference leading to cycle (through reference chain: me.wolfyscript.customcrafting.recipes.CraftingRecipeShaped["ingredients"]->java.util.ImmutableCollections$Map1["D"]->me.wolfyscript.customcrafting.recipes.items.Ingredient["items"]->java.util.ArrayList[0]->com.wolfyscript.utilities.bukkit.world.items.reference.StackReference["identifier"]->me.wolfyscript.utilities.compatibility.plugins.mmoitems.MMOItemsStackIdentifier["type"]->net.Indyuce.mmoitems.api.Type["supertype"])
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.exc.InvalidDefinitionException.from(InvalidDefinitionException.java:77)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.SerializerProvider.reportBadDefinition(SerializerProvider.java:1300)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.BeanPropertyWriter._handleSelfReference(BeanPropertyWriter.java:944)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:722)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:774)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:178)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:728)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:774)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeWithType(BeanSerializerBase.java:657)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:730)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:774)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:178)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serializeContents(IndexedListSerializer.java:119)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serialize(IndexedListSerializer.java:79)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serialize(IndexedListSerializer.java:18)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:728)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:774)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:178)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeFields(MapSerializer.java:808)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeWithoutTypeInfo(MapSerializer.java:764)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:720)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:35)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:728)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:774)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeWithType(BeanSerializerBase.java:657)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.impl.TypeWrappedSerializer.serialize(TypeWrappedSerializer.java:32)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:480)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:319)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ObjectWriter$Prefetch.serialize(ObjectWriter.java:1518)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ObjectWriter._writeValueAndClose(ObjectWriter.java:1219)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ObjectWriter.writeValue(ObjectWriter.java:1027)
[14:53:09 WARN]:        at customcrafting-spigot-4.16.9.1.jar//me.wolfyscript.customcrafting.handlers.LocalStorageLoader.save(LocalStorageLoader.java:367)
[14:53:09 WARN]:        at customcrafting-spigot-4.16.9.1.jar//me.wolfyscript.customcrafting.recipes.CustomRecipe.save(CustomRecipe.java:359)
[14:53:09 WARN]:        at customcrafting-spigot-4.16.9.1.jar//me.wolfyscript.customcrafting.recipes.CustomRecipe.save(CustomRecipe.java:347)
[14:53:09 WARN]:        at customcrafting-spigot-4.16.9.1.jar//me.wolfyscript.customcrafting.data.cache.recipe_creator.RecipeCache.save(RecipeCache.java:202)
[14:53:09 WARN]:        at customcrafting-spigot-4.16.9.1.jar//me.wolfyscript.customcrafting.gui.recipe_creator.ClusterRecipeCreator.lambda$registerSaveButtons$22(ClusterRecipeCreator.java:222)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.utilities.api.inventory.gui.GuiHandler.onChat(GuiHandler.java:522)
[14:53:09 WARN]:        at wolfyutils-spigot-4.17-beta.3.jar//com.wolfyscript.utilities.bukkit.commands.InputCommand.lambda$execute$1(InputCommand.java:42)
[14:53:09 WARN]:        at org.bukkit.craftbukkit.v1_20_R3.scheduler.CraftTask.run(CraftTask.java:101)
[14:53:09 WARN]:        at org.bukkit.craftbukkit.v1_20_R3.scheduler.CraftScheduler.mainThreadHeartbeat(CraftScheduler.java:482)
[14:53:09 WARN]:        at net.minecraft.server.MinecraftServer.b(MinecraftServer.java:1646)
[14:53:09 WARN]:        at net.minecraft.server.dedicated.DedicatedServer.b(DedicatedServer.java:447)
[14:53:09 WARN]:        at net.minecraft.server.MinecraftServer.a(MinecraftServer.java:1525)
[14:53:09 WARN]:        at net.minecraft.server.MinecraftServer.w(MinecraftServer.java:1226)
[14:53:09 WARN]:        at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:319)
[14:53:09 WARN]:        at java.base/java.lang.Thread.run(Thread.java:840)`

**Comment by WolfyScript** [comment]
> Is there any update on this issue? I updated CC to 4.16.9.1 + WolfyUtils 4.17-beta.3. MMOItems 6.10 Still can't create any recipe but with slightly different error.
> 
> `[14:53:09 WARN]: me.wolfyscript.lib.com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Direct self-reference leading to cycle (through reference chain: me.wolfyscript.customcrafting.recipes.CraftingRecipeShaped["ingredients"]->java.util.ImmutableCollections$Map1["D"]->me.wolfyscript.customcrafting.recipes.items.Ingredient["items"]->java.util.ArrayList[0]->com.wolfyscript.utilities.bukkit.world.items.reference.StackReference["identifier"]->me.wolfyscript.utilities.compatibility.plugins.mmoitems.MMOItemsStackIdentifier["type"]->net.Indyuce.mmoitems.api.Type["supertype"]) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.exc.InvalidDefinitionException.from(InvalidDefinitionException.java:77) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.SerializerProvider.reportBadDefinition(SerializerProvider.java:1300) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.BeanPropertyWriter._handleSelfReference(BeanPropertyWriter.java:944) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:722) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:774) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:178) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:728) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:774) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeWithType(BeanSerializerBase.java:657) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:730) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:774) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:178) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serializeContents(IndexedListSerializer.java:119) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serialize(IndexedListSerializer.java:79) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serialize(IndexedListSerializer.java:18) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:728) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:774) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:178) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeFields(MapSerializer.java:808) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeWithoutTypeInfo(MapSerializer.java:764) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:720) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:35) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:728) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:774) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeWithType(BeanSerializerBase.java:657) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.impl.TypeWrappedSerializer.serialize(TypeWrappedSerializer.java:32) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:480) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:319) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ObjectWriter$Prefetch.serialize(ObjectWriter.java:1518) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ObjectWriter._writeValueAndClose(ObjectWriter.java:1219) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.lib.com.fasterxml.jackson.databind.ObjectWriter.writeValue(ObjectWriter.java:1027) [14:53:09 WARN]: at customcrafting-spigot-4.16.9.1.jar//me.wolfyscript.customcrafting.handlers.LocalStorageLoader.save(LocalStorageLoader.java:367) [14:53:09 WARN]: at customcrafting-spigot-4.16.9.1.jar//me.wolfyscript.customcrafting.recipes.CustomRecipe.save(CustomRecipe.java:359) [14:53:09 WARN]: at customcrafting-spigot-4.16.9.1.jar//me.wolfyscript.customcrafting.recipes.CustomRecipe.save(CustomRecipe.java:347) [14:53:09 WARN]: at customcrafting-spigot-4.16.9.1.jar//me.wolfyscript.customcrafting.data.cache.recipe_creator.RecipeCache.save(RecipeCache.java:202) [14:53:09 WARN]: at customcrafting-spigot-4.16.9.1.jar//me.wolfyscript.customcrafting.gui.recipe_creator.ClusterRecipeCreator.lambda$registerSaveButtons$22(ClusterRecipeCreator.java:222) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//me.wolfyscript.utilities.api.inventory.gui.GuiHandler.onChat(GuiHandler.java:522) [14:53:09 WARN]: at wolfyutils-spigot-4.17-beta.3.jar//com.wolfyscript.utilities.bukkit.commands.InputCommand.lambda$execute$1(InputCommand.java:42) [14:53:09 WARN]: at org.bukkit.craftbukkit.v1_20_R3.scheduler.CraftTask.run(CraftTask.java:101) [14:53:09 WARN]: at org.bukkit.craftbukkit.v1_20_R3.scheduler.CraftScheduler.mainThreadHeartbeat(CraftScheduler.java:482) [14:53:09 WARN]: at net.minecraft.server.MinecraftServer.b(MinecraftServer.java:1646) [14:53:09 WARN]: at net.minecraft.server.dedicated.DedicatedServer.b(DedicatedServer.java:447) [14:53:09 WARN]: at net.minecraft.server.MinecraftServer.a(MinecraftServer.java:1525) [14:53:09 WARN]: at net.minecraft.server.MinecraftServer.w(MinecraftServer.java:1226) [14:53:09 WARN]: at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:319) [14:53:09 WARN]: at java.base/java.lang.Thread.run(Thread.java:840)`

Yeah, that is another issue, but it will be fixed in the next WolfyUtils beta soon.

**Comment by WolfyScript** [comment]
For the original loading issue, I implemented a sync loading option now. That should fix the issue.
Now I am contemplating if that sync loading option should be automatically enabled when it detects MMOItems.

**Comment by ShadySt** [comment]
I see some progress on this issue with WolfyUtils 4.17-beta.4 update! 
CC couldn't load old configs, so I created new from scratch. Now I can successfully create new recipes with MMOItems ingredients, but this problem returns after plugin reload or server restart. 
`[22:23:07 WARN]: [CustomCrafting] [LOCAL] [Error] Invalid Recipe: "customcrafting:fishing/rod_silver_dark_oak": GenerateLoreEvent may only be triggered synchronously. (through reference chain: me.wolfyscript.customcrafting.recipes.CraftingRecipeShaped["ingredients"]->java.util.LinkedHashMap["H"]->me.wolfyscript.customcrafting.recipes.items.Ingredient["items"]->java.util.ArrayList[0])
[22:23:07 WARN]: [CustomCrafting] [LOCAL] [Error]   Caused by: GenerateLoreEvent may only be triggered synchronously.`

**Comment by SinonCute** [comment]
same issue on WolfyUtils 4.17-beta.4

**Comment by WolfyScript** [comment]
I have uploaded a CustomCrafting update yesterday (and a fix just now), that includes the sync data loading from the CC beta.
Install CC 4.16.11 from here https://modrinth.com/plugin/customcrafting/version/4.16.11
and then before starting the server, edit the `config.yml` to enable:
```yml
data:
  sync: true
```

_It does not yet get enabled by default, because I couldn't find an intuitive way to do that just yet._

**Comment by YYYork** [comment]
issue are solved by this solution!

**Comment by ShadySt** [comment]
Yep, issues mentioned in this report are solved, great!
However, this update introduced new one, that needs a separate report, I suppose.
